fabric8-ui/fabric8-planner
# Refactored planner with Redux technology with (ngrx/store and ngrx/effects)

# ngrx update

**OSIO reference user story (ngrx)** - https://openshift.io/openshiftio/openshiftio/plan/detail/1910

## Goals
* Reduce the number of API calls.
* Use a centralized data store to save UI sate.
* Use reducer based technology to update a state and relay the new state through observables.
* Simplify the events with one directional flow.

## More details
* https://github.com/ngrx/platform
* [ngrx in planner presentation](https://docs.google.com/presentation/d/1DJnvKyMcKaSUFt12U6TxuD7l9BkZhoPRH_INrKIpU_U/edit#slide=id.g2f6d4a6735_0_171
)
## Changes
* All the components were migrated to **components_ngrx**. All the old components are still there for feature reference.
* **actions**, **effects** and **reducers** directories added for respective attributes.
* Most of the logics are moved to effects and reducers from components and services (old logics are still there and unused and will be cleaned up later).
* All the service responses are converted to UI models to reduce the memory use to the minimum.
* All the attributes have their respective UI models and service models and mappers that converts UI models to service models and vice versa.

## Test path
* Landing page redirects to Scenario by default
* Change any type-group and check the updates overall 
  * URL should be changed for the selected group type
  * Work item type list should update in quick add
  * List should refresh to match the items from current group type
* Select iteration from the side panel
  * URL should be changed for the selected iteration
  * Work item type list should update in quick add
  * List should refresh to match the items from current iteration
* Add iterations 
* Check show tree true and false for all the above.
* Quick add work item both from the top and inline.
* Check non-contextual detail page and quick preview.
* Try updating work items from both quick preview and detail page.
* Add comments, edit comments, try editing someone else's comment.
* Try applying filters to list 
* Try refreshing any URL and check if you see the same state everytime

# Drag and Drop update
**OSIO reference user story (drag and drop)** -
 https://openshift.io/openshiftio/openshiftio/plan/detail/1719

## Goals
* Re-order Work Items using Drag-and-Drop.
* All applied restrictions for re-ordering are in place.

## Changes
* Use `ng-drag-drop@4.0.0`: Drag and Drop Library for Angular
* Update `rh-ngx-datatable` to v12.0.9
